src/main/msp/msp.c

```diff
@@ -2318,20 +2381,121 @@ static mspResult_e mspProcessInCommand(uint8_t cmdMSP, sbuf_t *src)
 
             if (sbufBytesRemaining(src) >= 2) {
                 vtxSettingsConfigMutable()->power = sbufReadU8(src);
+                const uint8_t newPitmode = sbufReadU8(src);
                 if (vtxType != VTXDEV_UNKNOWN) {
                     // Delegate pitmode to vtx directly
-                    const uint8_t newPitmode = sbufReadU8(src);
                     unsigned vtxCurrentStatus;
                     vtxCommonGetStatus(vtxDevice, &vtxCurrentStatus);
                     if ((bool)(vtxCurrentStatus & VTX_STATUS_PIT_MODE) != (bool)newPitmode) {
                         vtxCommonSetPitMode(vtxDevice, newPitmode);
                     }
+                }
+            }
 
-                    if (sbufBytesRemaining(src)) {
-                        vtxSettingsConfigMutable()->lowPowerDisarm = sbufReadU8(src);
-                    }
+            if (sbufBytesRemaining(src)) {
+                    vtxSettingsConfigMutable()->lowPowerDisarm = sbufReadU8(src);
+            }
+
+            // API version 1.42
+            if (sbufBytesRemaining(src) >= 9) {
+                vtxSettingsConfigMutable()->pitModeFreq = sbufReadU16(src);
+#ifdef USE_VTX_TABLE
+                const uint8_t newBandCount = sbufReadU8(src);
+                const uint8_t newChannelCount = sbufReadU8(src);
+                const uint8_t newPowerCount = sbufReadU8(src);
+
+                if ((newBandCount > VTX_TABLE_MAX_BANDS) ||
+                    (newChannelCount > VTX_TABLE_MAX_CHANNELS) ||
+                    (newPowerCount > VTX_TABLE_MAX_POWER_LEVELS)) {
+                    return MSP_RESULT_ERROR;
+                }
+                vtxTableConfigMutable()->bands = newBandCount;
+                vtxTableConfigMutable()->channels = newChannelCount;
+                vtxTableConfigMutable()->powerLevels = newPowerCount;
+
+                // We expect that the detailed band/channel and power level messages will
+                // follow to repopulate the tables, so clear them in anticipation.
+                for (int i = 0; i < VTX_TABLE_MAX_BANDS; i++) {
+                    vtxTableConfigClearBand(vtxTableConfigMutable(), i);
+                    vtxTableConfigClearChannels(vtxTableConfigMutable(), i, 0);
+                }
+                vtxTableConfigClearPowerLabels(vtxTableConfigMutable(), 0);
+                vtxTableConfigClearPowerValues(vtxTableConfigMutable(), 0);
+#else
+                sbufReadU8();
+                sbufReadU8();
+                sbufReadU8();
+#endif
+                // Added standalone values for band, channel and frequency to move
+                // away from the flawed encoded combined method originally implemented.
```

Comment:
this is interesting, what's the plan for the old method (above)? do we expect to maintain it forever or should it be in the release notes that it will be removed in the next version and that MSP client authors (configurator, etc) need to update their code?  If we do plan to remove the implementation of setting band/channel/frequence code above then this approach of adding the new values at the end doesn't make it easy unless the solution is to still accept the bytes, but ignore them.

I have a sneaky feeling that this message, MSP_SET_VTX_CONFIG, is now doing too much.

Perhaps the VTX table stuff (band count, channel count, power count), should be in new `MSP_GET/SET_VTXTABLE_PARAMETERS` messages and that the changes below (setting vtx band/channel/freq/power/pitfreq/lowpowerdisarm) should perhaps be the only things in MSP_SET_VTX_CONFIG.

Also, if you just want to change frequency using the new method you'd have to resend the entire vtx band/channel/power levels as they will have been cleared.

Reply:
> Perhaps the VTX table stuff (band count, channel count, power count), should be in new MSP_GET/SET_VTXTABLE_PARAMETERS messages and that the other settings (vtx band/channel/freq/power/pitfreq/lowpowerdisarm) should perhaps be the only things in MSP_SET_VTX_CONFIG.
>
>Also, if you just want to change frequency using the new method you'd have to resend the entire vtx band/channel/power levels as they will have been cleared.

Yes this is a good point. We were trying to minimize the number of new MSP messages because there aren't that many usable IDs free. But the current logic of clearing the `vtxtable` on `MSP_SET_VTX_CONFIG` under the expectation that the other message will be sent is flawed since `MSP_SET_VTX_CONFIG` (although not used currently in the Configurator) is used from Lua. And Lua won't know (or have the ability) to send the bands/channels and power levels.

Reply:
> this is interesting, what's the plan for the old method (above)? do we expect to maintain it forever or should it be in the release notes that it will be removed in the next version and that MSP client authors (configurator, etc) need to update their code? If we do plan to remove the implementation of setting band/channel/frequence code above then this approach of adding the new values at the end doesn't make it easy unless the solution is to still accept the bytes, but ignore them.

Generally we don't change legacy MSP and the solution has always been to maintain the previous and add on for additional functionality. There are a number of messages that have "duplicate" newer versions of data elements that have different data types or other logic associated with them that wouldn't be supported by a consumer using an older API version.

At some point we need a "reset" and specify an API version that no longer provides backwards compatibility - but that's not now.

Reply:
> Also, if you just want to change frequency using the new method you'd have to resend the entire vtx band/channel/power levels as they will have been cleared.

As a suggestion for how to address this:
- if the MSP message received for `MSP_SET_VTX_CONFIG` does not include the fields added in API 1.42, then it should not be clearing the VTX tables - legacy clients won't know about them;
- for 'full' `MSP_SET_VTX_CONFIG` messages, we could include a boolean field `clearVtxTables` to control clearing / not clearing of VTX tables.

> At some point we need a "reset" and specify an API version that no longer provides backwards compatibility - but that's not now.

It is still my hope that we will be doing this as part of a move to using parameter group based messages to exchange configuration information, and generate the client stubs for them from their definitions.

Reply:
Good ideas on how to trigger (or not) the clearing of the vtxtable. I'll implement.

Reply:
i think that using a message to configure vtx table params is a good way to go.

less conditional logic than other proposed solutions, less testing required, no backwards compatibility issues, and no reasons to change the message other than for vtxtable reasons.

Reply:
I don't really think it's worthwhile to use two additional MSP messages (get and set) for this. The settings are describing the capabilities of the vtx so they make sense to be in the existing `MSP_VTX_CONFIG` and `MSP_SET_VTX_CONFIG`. We can handle the backwards compatibility issues just like we do with every other MSP message.